### PR TITLE
Only one target platform, only one native implementation file — `patch.cpp`

### PR DESCRIPTION
**Pull request by evgenykochetkov**
This is a prerequisite for #879 

- `xod-project`

  - no notion of different impl types
  - because there is only one kind of implementation left, no need to store them in `impls`, it can be extracted straight from attachments

  - API changes:
    - removed:
      - `listImpls`
      - `hasImpls`
      - `getImplByArray`

    - added:
      - `hasImpl :: Patch -> Boolean`
      - `setImpl :: Source -> Patch -> Patch`

    - changed signatures:
      - `getImpl` is now `Patch -> Maybe Source`
      - `hasImpl` is now `Patch -> Boolean`
      - `flatten` is now `Project -> PatchPath -> Either Error Project`

- `xod-fs`
  - does not know about implementation files — they all are just attachments

+ minor changes in other backages

### CODE REVIEW

File: packages/xod-fs/test/fixtures/workspace/__lib__/xod/core/pot/patch.cpp [hunk 1]
```diff
@@ -0,0 +1 @@
+// should be imported into impl.cpp
```
Comment (nkrkv) [inline packages/xod-fs/test/fixtures/workspace/__lib__/xod/core/pot/patch.cpp]: What’s this?

File: packages/xod-fs/test/load.spec.js [hunk 2]
```diff
@@ -99,28 +96,11 @@ describe('Loader', () => {
...
-          });
-        }).then(
-          implsLoaded => assert.equal(implsLoaded, true, 'some implementations were not loaded')
-        );
-    });
+    it('should return project without libs', () =>
+      Loader.loadProjectWithoutLibs(
+        path.resolve(workspace, './__lib__/xod/core')
+      )
+    );
```
Comment (nkrkv) [inline packages/xod-fs/test/load.spec.js]: Looks like it no longer tests anything

File: packages/xod-fs/test/fixtures/workspace/__lib__/xod/core/pot/patch.cpp (hunk 1 above)
Comment (evgenykochetkov) [reply packages/xod-fs/test/fixtures/workspace/__lib__/xod/core/pot/patch.cpp]: Oops. It was an outdated placeholder. Fixed!

File: packages/xod-project/src/attachment.js [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+  (filename, encoding, content) => ({
+    filename,
+    encoding,
+    content,
+  })
+);
+
+export const createImplAttachment = def(
+  'createImplAttachment :: Source -> Attachment',
+  content => ({
```
Comment (brusherru) [inline packages/xod-project/src/attachment.js]: Better reuse `createAttachment`

File: packages/xod-project/src/attachment.js [hunk 4]
```diff
@@ -0,0 +1,41 @@
+import R from 'ramda';
+import { def } from './types';
+import { IMPL_FILENAME } from './constants';
+
+export const createAttachment = def(
+  'createImplAttachment :: String -> String -> String -> Attachment',
```
Comment (brusherru) [inline packages/xod-project/src/attachment.js]: Wrong function name in the signature

File: packages/xod-project/src/flatten.js [hunk 5]
```diff
@@ -746,29 +752,31 @@ export const extractPatches = R.curry((project, leafPaths, prefix, boundValues,
...
-      Patch.upsertNodes(newNodes)
-    )(Patch.createPatch());
-  }
-)(patch));
+const convertPatch = def(
+  'convertPatch :: Project -> [Pair PatchPath Patch] -> Patch -> Patch',
+  (project, leafPatches, patch) =>
+    R.ifElse(
+      Patch.hasImpl,
+      R.identity,
```
Comment (brusherru) [inline packages/xod-project/src/flatten.js]: `R.ifElse` could be replaced with `R.unless` in this case.

File: packages/xod-project/src/attachment.js (hunk 4 above)
Comment (evgenykochetkov) [reply packages/xod-project/src/attachment.js]: Nice catch, fixed!

File: packages/xod-project/src/attachment.js (hunk 3 above)
Comment (evgenykochetkov) [reply packages/xod-project/src/attachment.js]: Done

File: packages/xod-project/src/flatten.js (hunk 5 above)
Comment (evgenykochetkov) [reply packages/xod-project/src/flatten.js]: Done

**Comment by evgenykochetkov** [comment]
@nkrkv @brusherru I refactored the test in `xod-func-tools` that was stopping the build, check it out please

**Comment by nkrkv** [comment]
Yeeee. Not a huge fan of wasting 100 ms to just wait… but in the scale of our whole test suite, it is OK.

**Pull request merged**